py2pack/parse.py

```diff
@@ -0,0 +1,207 @@
+    pkg_info_dict = {}
+    for key, value in pkg_info_items:
+        key = key.lower().replace('-', '_')
+        if key in {'requires_dist', 'provides_extra'}:
+            val = dict.setdefault(pkg_info_dict, key, [])
+            val.append(value)
+        elif key in {'classifier'}:
+            val = dict.setdefault(pkg_info_dict, key + 's', [])
+            val.append(value)
+        elif key in {'project_url'}:
+            key1, val = value.split(',', 1)
+            pkg_info_dict.setdefault(key + 's', {})[key1.strip()] = val.strip()
+        else:
+            pkg_info_dict[key] = value
+    return {'info': pkg_info_dict, 'urls': []}
+
+
+def pypi_json_file(file_path):
+    # parse pypi json file
+    json_file = open(file_path, 'r')
+    js = pypi_json_stream(json_file)
+    json_file.close()
+    return js
+
+
+def pypi_json_stream(json_stream):
+    # parse pypi json stream
+    js = json.load(json_stream)
+    if 'info' not in js:
+        js = {'info': js}
+    if 'urls' not in js:
+        js['urls'] = []
+    return js
+
+
+def _check_if_pypi_archive_file(path):
+    # check if archive is python source
+    return path.count('/') == 1 and basename(path) == 'PKG-INFO'
+
+
+def pypi_archive_file(file_path):
+    # try to extract metadata from tar archive
+    if tarfile.is_tarfile(file_path):
+        with tarfile.open(file_path, 'r') as archive:
+            for member in archive.getmembers():
+                if _check_if_pypi_archive_file(member.name):
+                    return pypi_text_stream(StringIO(archive.extractfile(member).read().decode()))
+    # try to extract metadata from zip archive
+    elif zipfile.is_zipfile(file_path):
+        with zipfile.ZipFile(file_path, 'r') as archive:
+            for member in archive.namelist():
+                if _check_if_pypi_archive_file(member):
+                    return pypi_text_stream(StringIO(archive.open(member).read().decode()))
+    else:
+        raise TypeError("Can not extract '%s'. Not a tar or zip file" % file_path)
+    raise KeyError('PKG-INFO not found on archive ' + file_path)
+
+
+def fetch_local_data(args):
+    # autodetect localfile name and type and parse
+    localfile = args.localfile
+    local = args.local
+    # set localarchive argument
+    args.localarchive = None
+    if not localfile and local:
+        try:
+            args.fetched_data = pypi_text_metaextract(args.name)
+            return True
+        except metadata.PackageNotFoundError:
+            localfile = join(f'{args.name}.egg-info', 'PKG-INFO')
+    if isfile(localfile):
+        try:
+            data = pypi_archive_file(localfile)
+            args.localarchive = localfile
+        except TypeError:
+            try:
+                data = pypi_json_file(localfile)
+            except json.decoder.JSONDecodeError:
+                data = pypi_text_file(localfile)
+        args.fetched_data = data
+        return True
+    return False
+
+
+def fix_data(args):
```

Comment:
I don’t like this kind of all-encompassing function with ambiguous functionality. Could not the cleaning of data be put to specific getters, so they return already clean data?

Also, if what you want is actually a constructor for `data["info"]`, then just say so. Is it?